### ISSUE TITLE
Add option to dropout during inference. 

### ISSUE THREAD

**Issue by BaxterEaves** [NervanaSystems/neon#266]
I would like to do inference *with* dropout because it gives you a measure of your model's uncertainty. See [Yarin Gal and Zoubin Ghahramani's paper on dropout as Bayesian approximation](https://arxiv.org/abs/1506.02142) or [Yarin's awesome blog post](http://mlg.eng.cam.ac.uk/yarin/blog_3d801aa532c1ce.html).

Right now `Model.get_outputs` sets `inference=True` in `layers.fprop`, which has the effect of ignoring the dropout layer during inference so you get a point estimate. The quick and dirty solution is to add and `inference` argument to `Model.get_outputs`, e.g.,

```python
def get_outputs(self, dataset, inference=True):
```
The problem with this (apart from being confusingly named) is that the `inference` affects the `BatchNorm` layers as well. I guess the solution is to add an `allow_dropout` argument to `Model.get_outputs` that propagates through `fprop` the same way that that `inference` does.

Thoughts? I'm happy to put a pull request through.

**Comment by djebm2** [comment]
Originally dropout has been designed specifically for training right ?

**Comment by BaxterEaves** [comment]
Yep. Using it during training is supposed to reduce overfitting. To use it to get an estimate of the variance in the output, you provide a single input and then generate a bunch of outputs while leaving dropout on.

**Comment by jennifermyers** [comment]
Sorry for the delay in response. It sounds interesting and we'd be very happy to get a PR from you.

**Comment by utkarshsimha** [comment]
Is this still open? I'd like to work on this! :)
@jennifermyers Could I make a PR if @BaxterEaves isn't working on it?

**Comment by jennifermyers** [comment]
Hi @utkarshsimha - absolutely! :-)

**Comment by BaxterEaves** [comment]
@utkarshsimha Go for it. I won't be submitting a PR.
